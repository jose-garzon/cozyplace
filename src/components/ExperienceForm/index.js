// ------------------------------ import libraries
import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

// ------------------------------ import components
import { Rate } from '../Rate';
import { Icon } from '../Icons';
import { Button } from '../../global-styles/Buttons';
import { InputText, InputDate } from '../../global-styles/Inputs';
import { useGetMousePosition } from '../../hooks/useGetMousePosition';

// ------------------------------ import styles and images
import {
  ExpCreationContainer,
  ExpForm,
  Rating,
  Dropdown,
  Pics,
  Inputs,
  MainInputs,
  SubmitButtons,
} from './styles';

// -------- import redux actions
import { createPost } from '../../utils/createPost';
import { useFetchData } from '../../hooks/useFetchData';
import {
  showIndicator,
  setIndicatorPosition,
} from '../../actions/experiencesActions';
import { setSelection } from '../../actions/userActions';

// ------------------------------------ COMPONENT ------------------------------------//
export const ExperienceForm = () => {
  const dispatch = useDispatch();
  const mousePosition = useGetMousePosition();

  const tags = useFetchData(
    'https://cozyplace.herokuapp.com/api/tag/',
  );
  const [rating, setrating] = useState(1);
  const { token, _id } = useSelector(
    (state) => state.userReducer.userData,
  );
  const [form, setForm] = useState({
    user: _id,
    country: 'Mexico',
    location: 'Cancun',
    duration: '30 Min',
    tags: '5f9b54d3217dd209f45ecc55',
  });
  const history = useHistory();

  const handleChange = (e) => {
    setForm({ ...form, [e.target.name]: e.target.value });
  };

  const handleFile = (e) => {
    setForm({
      ...form,
      [e.target.name]: URL.createObjectURL(e.target.files[0]),
    });
  };

  const handleSubmit = async (e) => {
    e.preventDefault();
    const response = await createPost(form, token);
    if (response.status === 200) {
      dispatch(
        showIndicator({
          status: true,
          message: 'Experience Successfully Created 🎉',
        }),
      );
      dispatch(setIndicatorPosition(mousePosition));
      dispatch(setSelection('favorites'));

      history.push('/user');
    } else if (response.status === 400) {
      dispatch(
        showIndicator({
          status: true,
          message: 'Missing data, fill in all the fields ❌',
        }),
      );
      dispatch(setIndicatorPosition(mousePosition));
    }
  };

  useEffect(() => {
    setForm({ ...form, rating });
  }, [rating]);

  return (
    <ExpCreationContainer>
      <h2>Create your experience</h2>
      <ExpForm onSubmit={handleSubmit}>
        <Inputs>
          <MainInputs>
            <InputText
              type="text"
              name="title"
              required="required"
              placeholder="Create Name of the Experience"
              onChange={handleChange}
            />
            <Dropdown name="country" onChange={handleChange}>
              <option value="Mexico">Mexico</option>
              <option value="Colombia">Colombia</option>
            </Dropdown>
            <Dropdown name="location" onChange={handleChange}>
              <option value="Mexico">Cancun</option>
              <option value="Mexico">Guadalajara</option>
              <option value="Mexico">Mexico City</option>
              <option value="Mexico">Monterrey</option>
              <option value="Mexico">Merida</option>
              <option value="Mexico">Oaxaca</option>
              <option value="Mexico">Puebla</option>
              <option value="Mexico">Queretaro</option>
              <option value="Medellin">Barranquilla</option>
              <option value="Bogota">Bogota</option>
              <option value="Medellin">Bucaramanga</option>
              <option value="Medellin">Cali</option>
              <option value="Medellin">Cartagena</option>
              <option value="Medellin">Medellin</option>
              <option value="Medellin">Santa Marta</option>
            </Dropdown>
            <Dropdown name="tags" onChange={handleChange}>
              {tags.map((tag) => (
                <option key={tag._id} value={tag._id}>
                  {tag.tagname}
                </option>
              ))}
            </Dropdown>
            <Dropdown name="duration" onChange={handleChange}>
              <option value="30 Min">30 Min</option>
              <option value="1 Hour">1 Hour</option>
              <option value="1.5 Hour">1.5 Hour</option>
              <option value="2 Hour">2 Hour</option>
              <option value="2.5 Hour">2.5 Hour</option>
              <option value="3 Hour">3 Hour</option>
              <option value="3.5 Hour">3.5 Hour</option>
              <option value="4 Hour">4 Hour</option>
              <option value="4.5 Hour">4.5 Hour</option>
              <option value="4.5 Hour">1 Day</option>
              <option value="4.5 Hour">2 Days</option>
              <option value="4.5 Hour">3 Days</option>
              <option value="4.5 Hour">4 Days</option>
            </Dropdown>
            <InputDate
              type="date"
              name="date"
              onChange={handleChange}
              required="required"
            />
            <Rating>
              <p>Rating Stars</p>
              <Rate setrating={setrating} rate={rating} />
            </Rating>
          </MainInputs>
          <Pics>
            <label htmlFor="image">
              {form.image ? (
                <img
                  src={form.image}
                  alt="this is your experience"
                  pic="true"
                />
              ) : (
                <Icon type="plus" />
              )}
              <input
                type="file"
                name="image"
                accept="image/jpg,png"
                id="image"
                onChange={handleFile}
              />
            </label>
          </Pics>
        </Inputs>
        <textarea
          name="description"
          maxLength="1500"
          placeholder="Create you description"
          onChange={handleChange}
          required="required"
        />
        <SubmitButtons>
          <Button name="cancel" onClick={() => history.push('/user')}>
            Cancelar
          </Button>
          <Button main type="submit">
            Crear
          </Button>
        </SubmitButtons>
      </ExpForm>
    </ExpCreationContainer>
  );
};
