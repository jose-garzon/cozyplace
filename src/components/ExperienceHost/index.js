// ------------------------------ import libraries
import React from 'react';
import { useSelector } from 'react-redux';

// ------------------------------ import components
import { HostCard } from '../HostCard';
import { HostTravel } from '../HostTravel';
import { Reviews } from '../Reviews';
import userIcon from '../../assets/images/userIcon.svg';

// ------------------------------ import styles and images
import { BlockExp, DescriptionExp, DetailsHost } from './style';

// ------------------------------------ COMPONENT ------------------------------------//

export const ExperienceHost = () => {
  const current = useSelector(
    (state) => state.experiencesReducer.current,
  );

  const {
    title,
    description,
    location,
    duration,
    tags,
    rating,
    user,
  } = current;
  const experienceUser =
    user?.length > 0 ? user[0]?.fullname : 'Jon Doe';

  function validatePostUserImage() {
    let image;
    if (user?.length === 0) {
      image = userIcon;
    } else if (user && user[0]?.image === '') {
      image = userIcon;
    } else {
      image = user && user[0]?.image;
    }
    return image;
  }
  const userImage = validatePostUserImage();

  return (
    <BlockExp>
      <h1>{title}</h1>
      <DescriptionExp>
        <DetailsHost>
          <HostCard
            userImg={userImage}
            name={experienceUser}
            rating={rating ? rating[0] : null}
          />
          <div>
            <h3>What you will do</h3>
            <p>{description}</p>
          </div>
        </DetailsHost>
        <article>
          <HostTravel
            location={location}
            duration={duration}
            tag={tags ? tags[0].tagname : null}
          />
        </article>
      </DescriptionExp>
    </BlockExp>
  );
};
