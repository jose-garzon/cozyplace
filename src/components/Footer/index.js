// ------------------------------ import libraries
import React from 'react';

// ------------------------------ import components
import { Icon } from '../Icons';

// ------------------------------ import styles and images
import logo from '../../assets/images/logo.svg';
import {
  StyledFooter,
  Description,
  Rights,
  ListsGroup,
  List,
} from './styles';

// ------------------------------------ COMPONENT ------------------------------------//
// this is the footer. It has a short description and the rights declaration.
// Also has the list of destinations and interests.
// Finally it has the social network links

export const Footer = () => {
  return (
    <StyledFooter>
      <Description>
        <Rights>
          <img src={logo} alt="logo cozy place" />
          <p>
            For decades travellers have reached for Lonely Planet
            books when looking to plan and execute their perfect trip,
            but now, they can also let Lonely Planet Experiences lead
            the way.
          </p>
          <p>LÂ©2020 Thousand Sunny. All rights reserved.</p>
        </Rights>
        <ListsGroup>
          <List>
            <li>destinations</li>
            <li>Mexico</li>
            <li>Colombia</li>
          </List>
          <List>
            <li>interests</li>
            <li>Adventure Travel</li>
            <li>Art and Culture</li>
            <li>Wildlife and Nature</li>
            <li>Family Holidays</li>
            <li>Food and Drinks</li>
          </List>
        </ListsGroup>
      </Description>
    </StyledFooter>
  );
};
