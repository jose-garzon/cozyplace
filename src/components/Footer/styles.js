// ------------------------------ import libraries
import styled from 'styled-components';

// ------------------------------ import variables
import {
  colors,
  spacing,
  shadows,
  border,
  fonts,
} from '../../global-styles/Variables';
import { media } from '../../global-styles/MediaQueries';

// ------------------------------------ COMPONENTS ------------------------------------//
export const StyledFooter = styled.footer`
  margin-top: ${spacing.space4};
  padding: ${spacing.space3};
  border-radius: ${border.borderRadiusBig};
  text-align: center;
  box-shadow: ${shadows.innerShadow};

  ${media.mobile} {
    margin: ${spacing.space2};
    padding: ${spacing.space2};
  }
`;

export const Description = styled.div`
  display: flex;
  flex-wrap: wrap;
  margin-bottom: ${spacing.space2};
  justify-content: space-between;
`;

export const Rights = styled.div`
  width: 40%;
  text-align: left;

  ${media.mobile} {
    width: 100%;
  }

  img {
    width: ${spacing.space4};
  }
`;

export const ListsGroup = styled.div`
  display: flex;
  ${media.mobile} {
    margin: 0 auto;
  }
`;

export const List = styled.ul`
  list-style: none;

  ${media.mobile} {
    padding: 0;
  }

  li {
    margin-bottom: ${spacing.space1};
  }

  li:first-child {
    margin-bottom: ${spacing.space3};
    color: transparent;
    text-transform: uppercase;
    font-weight: ${fonts.bold};
    ${colors.mainBackgroundColor};
    -webkit-background-clip: text;
  }
`;
