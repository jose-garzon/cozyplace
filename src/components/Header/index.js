// ------------------------------ import libraries
import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link, useHistory } from 'react-router-dom';

// ------------------------------ import styles and images
import {
  StyledHeader,
  Brand,
  Logo,
  LinkedLogo,
  Image,
  MenuProfile,
} from './styles';
import { Button } from '../../global-styles/Buttons';
import logo from '../../assets/images/logo.svg';
import userIcon from '../../assets/images/userIcon.svg';

// -------- import redux actions
import { openLogin, openSign } from '../../actions/userActions';

// ------------------------------------ COMPONENT ------------------------------------//
// This is the header. It change if the user is logged.
// In the landing page it shows the login and sign up buttons.
// When the user is logged it shows the user picture and a link to home page.

export const Header = () => {
  const dispatch = useDispatch();
  const history = useHistory();
  const { token, image } = useSelector(
    (state) => state.userReducer.userData,
  );
  const userImage = image === '' ? userIcon : image;

  // this functions open the modal
  const openLoginModal = () => dispatch(openLogin());
  const openSignModal = () => dispatch(openSign());
  const logout = () => {
    localStorage.clear();
    window.location.reload();
  };

  return (
    <StyledHeader>
      <Brand>
        <LinkedLogo to="/">
          <Logo src={logo} alt="Cozy place logo" />
        </LinkedLogo>
      </Brand>
      {token ? (
        history.location.pathname === '/user' ? null : (
          <MenuProfile>
            <Link to="/user">
              <Image src={userImage} alt="" />
            </Link>
            <ul>
              <Button onClick={logout}>logout</Button>
            </ul>
          </MenuProfile>
        )
      ) : (
        <div>
          <Button onClick={openLoginModal}>login</Button>
          {window.innerWidth <= 400 ? null : (
            <Button main onClick={openSignModal}>
              Sign up
            </Button>
          )}
        </div>
      )}
    </StyledHeader>
  );
};
