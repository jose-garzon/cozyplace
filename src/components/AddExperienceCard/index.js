// ------------------------------ import libraries
import React from 'react';
import { useHistory } from 'react-router-dom';

// ------------------------------ import components
import { Icon } from '../Icons';

// ------------------------------ import styles and images
import { CardStyled } from './styles';

// ------------------------------------ COMPONENT ------------------------------------//

export const AddExperienceCard = () => {
  const history = useHistory();

  function handleClick() {
    history.push('/create-experience');
  }
  return (
    <CardStyled>
      <h3>Add an experience</h3>
      <Icon type="plus" click={handleClick} />
    </CardStyled>
  );
};
