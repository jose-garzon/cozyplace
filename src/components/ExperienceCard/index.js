// ------------------------------ import libraries
import React, { useState } from 'react';
import { Link } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';

// ------------------------------ import components
import { Rate } from '../Rate';
import { Icon } from '../Icons';
import { addFavorite } from '../../utils/addFavorite';
import { deleteFavorite } from '../../utils/deleteFavorite';

// ------------------------------ import styles and images
import { CardStyled, CardImg, TitleCard } from './styles';
import mockupPicture from '../../assets/images/mockupPicture.jpg';
import { useGetMousePosition } from '../../hooks/useGetMousePosition';

// -----------------Redux actions
import { getUserData } from '../../actions/userActions';
import {
  showIndicator,
  setIndicatorPosition,
} from '../../actions/experiencesActions';

// ------------------------------------ COMPONENT ------------------------------------//
// this card shows the experiences, their rating and name.
// it is in diferent pages, so it's styles has many validations to the difernts positions in the diferentes grids

export const ExperienceCard = ({
  position,
  page,
  _id,
  title,
  rating,
  image,
  inUser,
}) => {
  const dispatch = useDispatch();
  const userData = useSelector((state) => state.userReducer.userData);
  const { token, favorite } = userData;
  const mousePosition = useGetMousePosition();

  const findInFavList = favorite?.find(
    (experience) => experience._id === _id,
  );
  const isFav = favorite?.includes(findInFavList);
  const [liked, setLiked] = useState(isFav);
  const displayHeartBroken = token && isFav && inUser;
  const displayHeart = token && !inUser;

  async function addFav() {
    const response = await addFavorite(_id, token);
    if (response.status === 200) {
      setLiked(true);
      dispatch(getUserData(userData._id, token));
      dispatch(
        showIndicator({
          status: true,
          message: 'Experience added to favorite! üéâ',
        }),
      );
      dispatch(setIndicatorPosition(mousePosition));
    }
  }

  async function deleteFav() {
    const response = await deleteFavorite(_id, token);
    if (response.status === 200) {
      setLiked(false);
      dispatch(getUserData(userData._id, token));
      dispatch(
        showIndicator({
          status: true,
          message: 'Experience deleted from favorite ‚ùå',
        }),
      );
      dispatch(setIndicatorPosition(mousePosition));
    }
  }

  return (
    <CardStyled position={`${page}${position}`} key={_id}>
      <Link to={`/experience/${_id}`}>
        <Rate rate={rating[0]} />
        <div>
          <CardImg src={image[0]} alt={title} />
          <TitleCard>{title}</TitleCard>
        </div>
      </Link>
      {displayHeartBroken ? (
        <Icon type="heartBroken" click={deleteFav} />
      ) : null}
      {displayHeart ? (
        <Icon
          type={liked ? 'favoriteHeart' : 'emptyHeart'}
          click={liked ? deleteFav : addFav}
        />
      ) : null}
    </CardStyled>
  );
};
