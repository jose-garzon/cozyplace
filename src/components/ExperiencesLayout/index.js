// ------------------------------ import libraries
import React, { useLayoutEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

// ------------------------------ import components
import { ExperienceCard } from '../ExperienceCard';
import { LoadingCard } from '../LoadingCard';
import { NotFoundSearch } from '../NotFoundSearch';

// ------------------------------ import styles and images
import { MainStyled } from './styles';

// -------- import redux actions
import { getAllExperiences } from '../../actions/experiencesActions';

// ------------------------------------ COMPONENT ------------------------------------//
// this is the layout to show the experiences in home page.

export const ExperiencesLayout = () => {
  const dispatch = useDispatch();
  const { searchExperiences, loading } = useSelector(
    (state) => state.experiencesReducer,
  );

  useLayoutEffect(() => {
    dispatch(getAllExperiences());
  }, []);

  const notFound =
    loading === false && searchExperiences.length === 0;

  return (
    <MainStyled>
      {loading
        ? [...Array(10)].map((card, i = 0) => {
            i++;
            return <LoadingCard key={i} position={i} page="home" />;
          })
        : searchExperiences.slice(0, 10).map((experience, i = 0) => {
            i++;
            return (
              <ExperienceCard
                position={i}
                page="home"
                key={experience._id}
                {...experience}
              />
            );
          })}
      {notFound ? <NotFoundSearch page="home" /> : null}
    </MainStyled>
  );
};
