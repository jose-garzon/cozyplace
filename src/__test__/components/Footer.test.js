// ------------------------------ import libraries
import React from 'react';
import { create } from 'react-test-renderer';
import { mount } from 'enzyme';

// ------------------------------ import components
import { Footer } from '../../components/Footer';

// ------------------------------------ TESTS ------------------------------------//
describe('<Footer />', () => {
  test('Render Footer component', () => {
    const footer = mount(<Footer />);
    expect(footer.length).toEqual(1);
  });
  test('Footer Snapshot', () => {
    const footer = create(<Footer />);
    expect(footer.toJSON()).toMatchSnapshot();
  });
});
