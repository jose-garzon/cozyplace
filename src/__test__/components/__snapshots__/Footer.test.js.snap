// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> Footer Snapshot 1`] = `
<footer
  className="styles__StyledFooter-sc-1569jkl-0 ZOThD"
>
  <div
    className="styles__Description-sc-1569jkl-1 buvuhc"
  >
    <div
      className="styles__Rights-sc-1569jkl-2 cpNfsx"
    >
      <img
        alt="logo cozy place"
        src={Object {}}
      />
      <p>
        For decades travellers have reached for Lonely Planet books when looking to plan and execute their perfect trip, but now, they can also let Lonely Planet Experiences lead the way.
      </p>
      <p>
        LÂ©2020 Thousand Sunny. All rights reserved.
      </p>
    </div>
    <div
      className="styles__ListsGroup-sc-1569jkl-3 zVJdj"
    >
      <ul
        className="styles__List-sc-1569jkl-4 dvbFSF"
      >
        <li>
          destinations
        </li>
        <li>
          Mexico
        </li>
        <li>
          Colombia
        </li>
      </ul>
      <ul
        className="styles__List-sc-1569jkl-4 dvbFSF"
      >
        <li>
          interests
        </li>
        <li>
          Adventure Travel
        </li>
        <li>
          Art and Culture
        </li>
        <li>
          Wildlife and Nature
        </li>
        <li>
          Family Holidays
        </li>
        <li>
          Food and Drinks
        </li>
      </ul>
    </div>
  </div>
  <div>
    <img
      alt="facebook"
      className="styles__StyledIcon-urxmhz-0 bOQGgD"
      src={Object {}}
      tabIndex="0"
      type="image"
    />
    <img
      alt="instagram"
      className="styles__StyledIcon-urxmhz-0 bOQGgD"
      src={Object {}}
      tabIndex="0"
      type="image"
    />
    <img
      alt="youtube"
      className="styles__StyledIcon-urxmhz-0 bOQGgD"
      src={Object {}}
      tabIndex="0"
      type="image"
    />
    <img
      alt="twitter"
      className="styles__StyledIcon-urxmhz-0 bOQGgD"
      src={Object {}}
      tabIndex="0"
      type="image"
    />
  </div>
</footer>
`;
